BinaryEncoding.md

```diff
@@ -119,11 +119,15 @@ The type section declares all function signatures that will be used in the modul
 | entries | `type_entry*` | repeated type entries as described below |
 
 #### Type entry
-| Field | Type | Description |
+| Field | Type/Value | Description |
 | ----- |  ----- | ----- |
+| constructor | `0x05` | the function type constructor |
 | param_count | `varuint32` | the number of parameters to the function |
-| return_type | `value_type?` | the return type of the function, with `0` indicating no return type |
 | param_types | `value_type*` | the parameter types of the function |
+| return_count | `uint8` | the number of results from the function (0 or 1) |
+| return_type | `value_type?` | the result type of the function (if return_count is 1) |
```

Comment:
On 5 April 2016 at 15:23, JSStats <notifications@github.com> wrote:

> In BinaryEncoding.md
> <https://github.com/WebAssembly/design/pull/636#discussion_r58536107>:
>
> >  | param_types | `value_type*` | the parameter types of the function |
> > +| return_count | `uint8` | the number of results from the function (0 or 1) |
> > +| return_type | `value_type?` | the result type of the function (if return_count is 1) |
>
> Just as with sections, a string might be needed for each class of type
> declaration to avoid name clashes, and a size will be needed to skip
> unknown type declarations.
>

Unlike for unknown sections, which an implementation can ignore, there is
no sane way to handle unknown types, so that's not useful here.


> Might it be prudent to just define the return_count as a varuint32?
> Could the return_type be defined a having return_count repeated value_type
> entries?
>

Note that uint8 is forward compatible with varuint32, and so is the
treatment of the count. That is intentional, of course.

The 0 and 1 return_count limitation might not even be a limit for some
> tools. For example a tool to convert to the text format need not be limited
> to just 0 and 1, so perhaps the limit should be noted elsewhere, perhaps in
> validation rules for the MVP, and not a property of the binary encoding?
>
That would extend the AST, though. In particular, it would introduce
malformed types into the AST, which we currently don't have.

This is adding one byte per function that seems unnecessary and elsewhere
> some people have objected to this.
>

No, this is one byte per different function _type_, typically a much
smaller number.

The function signatures are ordered, in the same order as the function
> bodies are defined, and the count defines the number of functions which
> seems useful for loading a vector etc. The count now would be the number of
> type declarations.
>
I don't follow.


> Is there any technical reason for not just using separate sections for
> future classes of type declarations?
>
Yes: then you would need distinguished ways to reference the different
kinds of types, in places where either would be allowed. I.e., you'd put
the cost on each use site instead of the fewer definition sites.